File: java/json-transform/src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionMathTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

public class TransformerFunctionMathTest extends BaseTest {
    @Test
    void inline() {
        var arr = new BigDecimal[] {new BigDecimal(4), new BigDecimal(2)};
        assertTransformation(arr, "$$math(+,$[0],$[1])", fromJson("6"));
        assertTransformation(arr, "$$math($[0],+,$[1])", fromJson("6"));
        assertTransformation(arr, "$$math($[0],+)", arr[0]);
        assertTransformation(arr, "$$math($[0],-,$[1])", fromJson("2"));
        assertTransformation(arr, "$$math($[0],*,$[1])", fromJson("8"));
        assertTransformation(arr, "$$math($[0],/,$[1])", fromJson("2"));
        assertTransformation(arr, "$$math($[0],//,3)", fromJson("1"));
        assertTransformation(arr, "$$math($[0],%,3)", fromJson("1"));
        assertTransformation(arr, "$$math($[1],^,3)", fromJson("8"));
        assertTransformation(arr, "$$math(3,&,1)", fromJson("1"));
        assertTransformation(arr, "$$math(6,AND,3)", fromJson("2"));
        assertTransformation(arr, "$$math(6,|,3)", fromJson("7"));
        assertTransformation(arr, "$$math(6,~,3)", fromJson("5"));
        assertTransformation(arr, "$$math(6,>>,1)", fromJson("3"));
        assertTransformation(arr, "$$math(6,<<,3)", fromJson("48"));
        assertTransformation(arr, "$$math(MIN,$[0],$[1])", fromJson("2"));
        assertTransformation(arr, "$$math(MAX,$[0],$[1])", fromJson("4"));
        assertTransformation(arr, "$$math(SQRT,81)", fromJson("9"));
        assertTransformation(arr, "$$math(SQRT):81", fromJson("9"));
        assertTransformation(arr, "$$math(SQRT):$$math($[0],^,2)", arr[0]);
        assertTransformation(arr, "$$math(ROUND,4.6)", fromJson("5"));
        assertTransformation(arr, "$$math(ROUND):4.6", fromJson("5"));
        assertTransformation(arr, "$$math(ROUND,1):4.66", fromJson("4.7"));
        assertTransformation(arr, "$$math(ROUND,4.66,1)", fromJson("4.7"));
        assertTransformation(arr, "$$math(4.66,ROUND,1)", fromJson("4.7"));
        assertTransformation(arr, "$$math(FLOOR,4.6)", fromJson("4"));
        assertTransformation(arr, "$$math(FLOOR):4.6", fromJson("4"));
        assertTransformation(arr, "$$math(FLOOR,1):4.66", fromJson("4.6"));
        assertTransformation(arr, "$$math(CEIL,4.2)", fromJson("5"));
        assertTransformation(arr, "$$math(CEIL):4.2", fromJson("5"));
        assertTransformation(arr, "$$math(CEIL,1):4.22", fromJson("4.3"));
        assertTransformation(arr, "$$math(ABS,-10)", fromJson("10"));
        assertTransformation(arr, "$$math(ABS):-10", fromJson("10"));
        assertTransformation(arr, "$$math(NEG,$[0])", fromJson("-4"));
        assertTransformation(arr, "$$math(NEG):$[0]", fromJson("-4"));
        assertTransformation(arr, "$$math(SIG):42", fromJson("1"));
        assertTransformation(arr, "$$math(SIGNUM):-42", fromJson("-1"));
        assertTransformation(arr, "$$math(SIG):0", fromJson("0"));
    }

    @Test
    void object() {
        var arr = new BigDecimal[] {new BigDecimal(4), new BigDecimal(2)};
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "+", "$[0]", "$[1]" ]
}
"""), fromJson("6"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "$[0]", "+", "$[1]" ]
}
"""), fromJson("6"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "-", "$[0]", "$[1]" ]
}
"""), fromJson("2"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "*", "$[0]", "$[1]" ]
}
"""), fromJson("8"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "/", "$[0]", "$[1]" ]
}
"""), fromJson("2"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "//", "$[0]", "3" ]
}
"""), fromJson("1"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "$[0]", "//", 3 ]
}
"""), fromJson("1"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "%", "$[0]", "3" ]
}
"""), fromJson("1"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "^", "$[1]", "3" ]
}
"""), fromJson("8"));

        assertTransformation(arr, fromJson("""
{ "$$math": [ 3, "&", 1 ] }"""), fromJson("1"));
        assertTransformation(arr, fromJson("""
{ "$$math": [ 6, "&", 3 ] }"""), fromJson("2"));
        assertTransformation(arr, fromJson("""
{ "$$math": [ 6, "|", 3 ] }"""), fromJson("7"));
        assertTransformation(arr, fromJson("""
{ "$$math": [ 6, "XOR", 3 ] }"""), fromJson("5"));
        assertTransformation(arr, fromJson("""
{ "$$math": [ 6, ">>", 1 ] }"""), fromJson("3"));
        assertTransformation(arr, fromJson("""
{ "$$math": [ 6, "<<", 3 ] }"""), fromJson("48"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "sqrt", 81 ]
}
"""), fromJson("9"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "min", "$[0]", "$[1]" ]
}
"""), fromJson("2"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "max", "$[0]", "$[1]" ]
}
"""), fromJson("4"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "round", 4.6 ]
}
"""), fromJson("5"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "round", 4.66, 1 ]
}
"""), fromJson("4.7"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "floor", 4.6 ]
}
"""), fromJson("4"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "floor", 4.66, 1 ]
}
"""), fromJson("4.6"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "ceil", 4.2 ]
}
"""), fromJson("5"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "ceil", 4.22, 1 ]
}
"""), fromJson("4.3"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "abs", -10 ]
}
"""), fromJson("10"));
        assertTransformation(arr, fromJson("""
{
  "$$math": [ "neg", "$[0]" ]
}
"""), fromJson("-4"));
        assertTransformation(fromJson("[\"abs\",-10]"), fromJson("""
{
  "$$math": "$"
}
"""), fromJson("10"));

        assertTransformation(arr, fromJson("""
{ "$$math": [ "SIG", 42 ] }"""), fromJson("1"));
        assertTransformation(arr, fromJson("""
{ "$$math": [ "SIGNUM", -42 ] }"""), fromJson("-1"));
        assertTransformation(arr, fromJson("""
{ "$$math": [ "SIG", 0 ] }"""), fromJson("0"));

    }

    @Test
    void combineScaling() {
        assertTransformation(null, "$$decimal(2):$$math(1,*,0.987654321)", fromJson("0.99"));
        assertTransformation(null, "$$decimal(2,FLOOR):$$math(1,*,0.987654321)", fromJson("0.98"));
    }
}

File: javascript/json-transform/src/__tests__/functions/TransformerFunctionMath.test.ts
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";
import { BigDecimal } from "../../functions/common/FunctionHelpers";

describe("TransformerFunctionMath", () => {
  test("inline", async () => {
    const arr = [BigDecimal(4), BigDecimal(2)];
    await assertTransformation(arr, "$$math(+,$[0],$[1])", BigDecimal(6));
    await assertTransformation(arr, "$$math($[0],+,$[1])", BigDecimal(6));
    await assertTransformation(arr, "$$math($[0],+)", arr[0]);
    await assertTransformation(arr, "$$math($[0],-,$[1])", BigDecimal(2));
    await assertTransformation(arr, "$$math($[0],*,$[1])", BigDecimal(8));
    await assertTransformation(arr, "$$math($[0],/,$[1])", BigDecimal(2));
    await assertTransformation(arr, "$$math($[0],//,3)", BigDecimal(1));
    await assertTransformation(arr, "$$math($[0],%,3)", BigDecimal(1));
    await assertTransformation(arr, "$$math($[1],^,3)", BigDecimal(8));
    await assertTransformation(arr, "$$math(3,&,1)", BigDecimal(1));
    await assertTransformation(arr, "$$math(6,AND,3)", BigDecimal(2));
    await assertTransformation(arr, "$$math(6,|,3)", BigDecimal(7));
    await assertTransformation(arr, "$$math(6,~,3)", BigDecimal(5));
    await assertTransformation(arr, "$$math(6,>>,1)", BigDecimal(3));
    await assertTransformation(arr, "$$math(6,<<,3)", BigDecimal(48));
    await assertTransformation(arr, "$$math(MIN,$[0],$[1])", BigDecimal(2));
    await assertTransformation(arr, "$$math(MAX,$[0],$[1])", BigDecimal(4));
    await assertTransformation(arr, "$$math(SQRT,81)", BigDecimal(9));
    await assertTransformation(arr, "$$math(SQRT):81", BigDecimal(9));
    await assertTransformation(arr, "$$math(SQRT):$$math($[0],^,2)", arr[0]);
    await assertTransformation(arr, "$$math(ROUND,4.6)", BigDecimal(5));
    await assertTransformation(arr, "$$math(ROUND):4.6", BigDecimal(5));
    await assertTransformation(arr, "$$math(ROUND,1):4.66", BigDecimal(4.7));
    await assertTransformation(arr, "$$math(ROUND,4.66,1)", BigDecimal(4.7));
    await assertTransformation(arr, "$$math(4.66,ROUND,1)", BigDecimal(4.7));
    await assertTransformation(arr, "$$math(FLOOR,4.6)", BigDecimal(4));
    await assertTransformation(arr, "$$math(FLOOR):4.6", BigDecimal(4));
    await assertTransformation(arr, "$$math(FLOOR,1):4.66", BigDecimal(4.6));
    await assertTransformation(arr, "$$math(CEIL,4.2)", BigDecimal(5));
    await assertTransformation(arr, "$$math(CEIL):4.2", BigDecimal(5));
    await assertTransformation(arr, "$$math(CEIL,1):4.22", BigDecimal(4.3));
    await assertTransformation(arr, "$$math(ABS,-10)", BigDecimal(10));
    await assertTransformation(arr, "$$math(ABS):-10", BigDecimal(10));
    await assertTransformation(arr, "$$math(NEG,$[0])", BigDecimal(-4));
    await assertTransformation(arr, "$$math(NEG):$[0]", BigDecimal(-4));
    await assertTransformation(arr, "$$math(SIG):42", BigDecimal(1));
    await assertTransformation(arr, "$$math(SIGNUM):-42", BigDecimal(-1));
    await assertTransformation(arr, "$$math(SIG):0", BigDecimal(0));
  });

  test("object", async () => {
    const arr = [BigDecimal(4), new BigDecimal(2)];
    await assertTransformation(
      arr,
      {
        $$math: ["+", "$[0]", "$[1]"],
      },
      BigDecimal(6),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["$[0]", "+", "$[1]"],
      },
      BigDecimal(6),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["-", "$[0]", "$[1]"],
      },
      BigDecimal(2),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["*", "$[0]", "$[1]"],
      },
      BigDecimal(8),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["/", "$[0]", "$[1]"],
      },
      BigDecimal(2),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["//", "$[0]", "3"],
      },
      BigDecimal(1),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["$[0]", "//", 3],
      },
      BigDecimal(1),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["%", "$[0]", "3"],
      },
      BigDecimal(1),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["^", "$[1]", "3"],
      },
      BigDecimal(8),
    );

    await assertTransformation(arr, { $$math: [3, "&", 1] }, BigDecimal(1));
    await assertTransformation(arr, { $$math: [6, "&", 3] }, BigDecimal(2));
    await assertTransformation(arr, { $$math: [6, "|", 3] }, BigDecimal(7));
    await assertTransformation(arr, { $$math: [6, "XOR", 3] }, BigDecimal(5));
    await assertTransformation(arr, { $$math: [6, ">>", 1] }, BigDecimal(3));
    await assertTransformation(arr, { $$math: [6, "<<", 3] }, BigDecimal(48));
    await assertTransformation(
      arr,
      {
        $$math: ["sqrt", 81],
      },
      BigDecimal(9),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["min", "$[0]", "$[1]"],
      },
      BigDecimal(2),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["max", "$[0]", "$[1]"],
      },
      BigDecimal(4),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["round", 4.6],
      },
      BigDecimal(5),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["round", 4.66, 1],
      },
      BigDecimal(4.7),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["floor", 4.6],
      },
      BigDecimal(4),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["floor", 4.66, 1],
      },
      BigDecimal(4.6),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["ceil", 4.2],
      },
      BigDecimal(5),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["ceil", 4.22, 1],
      },
      BigDecimal(4.3),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["abs", -10],
      },
      BigDecimal(10),
    );
    await assertTransformation(
      arr,
      {
        $$math: ["neg", "$[0]"],
      },
      BigDecimal(-4),
    );
    await assertTransformation(
      ["abs", -10],
      {
        $$math: "$",
      },
      BigDecimal(10),
    );

    await assertTransformation(arr, { $$math: ["SIG", 42] }, BigDecimal(1));
    await assertTransformation(arr, { $$math: ["SIGNUM", -42] }, BigDecimal(-1));
    await assertTransformation(arr, { $$math: ["SIG", 0] }, BigDecimal(0));
  });

  test("combineScaling", async () => {
    await assertTransformation(null, "$$decimal(2):$$math(1,*,0.987654321)", BigDecimal(0.99));
    await assertTransformation(null, "$$decimal(2,FLOOR):$$math(1,*,0.987654321)", BigDecimal(0.98));
  });
});
